clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -219,7 +219,7 @@ private long listOffset(TopicPartition partition, long timestamp) {
             for (PartitionRecords<K, V> part : this.records) {
                 Long consumed = subscriptions.consumed(part.partition);
                 if (this.subscriptions.assignedPartitions().contains(part.partition)
-                    && (consumed == null || part.fetchOffset == consumed)) {
```

Comment:
I might be mistaken, but I think the rebalance might still occur in between the time the records have been fetched and when they are returned to the user. This is one of the reasons why this code is tricky to work with; basically every time we finish a call to NetworkClient.poll(), we have to recheck the state of the world.